Drop script_pub_key arg from createNewBlock

Sjors opened the pull request:
Providing a script for the coinbase transaction is only done in test code and for (unoptimized) CPU solo mining.

Production miners use the `getblocktemplate` RPC which omits the coinbase transaction entirely from its block template, leaving it to external (pool) software to construct it.

This commit removes the `script_pub_key argument` from `createNewBlock()` in the Mining interface.

A coinbase script can still be passed via `BlockCreateOptions` instead. Tests are modified to do so.

Sjors commented on src/node/types.h:
```diff
@@ -43,6 +44,10 @@ struct BlockCreateOptions {
+    /**
+     * Script to put in the coinbase transaction.
+     */
+    std::optional<CScript> coinbase_script{};
```
In production this is always a dummy. `CreateNewBlock()` sets the script to `OP_TRUE` and then the `getblocktemplate` RPC omits the whole coinbase transaction.

It's not safe for the caller to set an _arbitrary_ dummy, because IIUC if it's larger than `coinbase_max_additional_weight` the `TestBlockValidity` internal check for `getblocktemplate` will spuriously fail. Which is why I think it's better for `CreateNewBlock` to decide what the dummy should be.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/31318#discussion_r1846813317

Would it make sense to declare the field as `CScript coinbase_output_script{CScript{} << OP_TRUE}` instead of `std::optional<CScript> coinbase_output_script{};` just to be clearer about what the default value is? Or maybe just add documentation to say what happens if this is not set?

I think the drawback of using std::optional here is that it's not really clear what happens if the option is not set. It's also not clear when you would want to set the option here, if ever.

Sjors replied:
I expanded the comment to explain things a bit more.

I think you should not set this option unless you care what the coinbase output is. For example the `block_assemble` benchmark uses `P2WSH_OP_TRUE`. I'm guessing that's because it wants to look at the performance related to spending from a segwit coinbase.

Sjors commented:
Renamed to `coinbase_output_script`, added warning and discouragement, fixed fuzz binary build.

TheCharlatan commented:
I'm not sure why this is really an improvement. Can you elaborate on why no external user would want to pass in a script pubkey directly, instead of manually manipulating the coinbase in the block template?

Sjors commented:
Afaik the only way external users make blocks today is the `getblocktemplate` RPC, which drops the coinbase transaction from the template it returns.

Similarly in `submitSolution`, used by Stratum v2, we expect to receive the full coinbase transaction and replace whatever was in the `CBlockTemplate`.

Unless you're solo mining, the pool is going to add its own outputs, manipulate the extra nonce, add its own name, etc. Even with a solo pool (a weird concept) you would configure the payout address in that software, not in the node.

There is [a way](https://github.com/stratum-mining/sv2-spec/blob/main/07-Template-Distribution-Protocol.md#72-newtemplate-server---client) in Stratum v2 to specify a list of outputs that you want in the coinbase. In my current implementation we use this to include the witness commitment, but nothing else.

In the longer run we could expand the interface to specify more outputs.

The special case for that is a single payout address that takes all the money from the block That could be used for solo mining with a single small ASIC that doesn't' need to grind the extra nonce (up to about 100 GH/s with [BIP320](https://github.com/bitcoin/bips/blob/master/bip-0320.mediawiki)). But that's fairly niche use case. As soon as you use Stratum v1 / v2 software you typically pass a payout address to the the pool software, not to the node.

ryanofsky reviewed (approved):
Concept ACK . It seems like a nice improvement to not require specifying an option that won't typically be used. But I'm a little hazy on the details of this. I think better documentation could help. Left a suggestion and some questions below.

Sjors commented:
Rebased to avoid spurious CI failure.

I'll take a look later if for any of the tests I can drop their custom coinbase output.

TheCharlatan commented:
Re https://github.com/bitcoin/bitcoin/pull/31318#issuecomment-2485792072

Thank you for spelling this out @Sjors.

Concept ACK

ryanofsky commented on src/node/miner.cpp:
```diff
@@ -120,6 +120,12 @@ std::unique_ptr<CBlockTemplate> BlockAssembler::CreateNewBlock(const CScript& sc
+    // Use dummy coinbase output script if none is provided
+    CScript coinbase_output_script{CScript() << OP_TRUE};
+    if (m_options.coinbase_output_script) {
+        coinbase_output_script = m_options.coinbase_output_script.value();
```
In commit "Drop script_pub_key arg from createNewBlock" (42f880aeb080070a42229d0bccd517100a3fe7fb)

Would suggest using '*' instead of `.value()` since this code isn't intending to trigger the `std::bad_optional_access` exception, and to make it more obvious a dereference is happening.

Sjors replied:
I changed `coinbase_output_script` to no longer be `std::optional`.

ryanofsky commented on src/node/miner.cpp:
```diff
@@ -106,7 +106,7 @@ void BlockAssembler::resetBlock()
 }
 
-std::unique_ptr<CBlockTemplate> BlockAssembler::CreateNewBlock(const CScript& scriptPubKeyIn)
+std::unique_ptr<CBlockTemplate> BlockAssembler::CreateNewBlock()
```
In commit "Drop script_pub_key arg from createNewBlock" (42f880aeb080070a42229d0bccd517100a3fe7fb)

Would it make sense to change the `BlockAssembler::CreateNewBlock()` signature in a separate followup commit? The test changes in this commit vastly outnumber the non-test changes, which I think make it more likely a bug in the non-test code could be missed.

Maybe this there could be two overloads initially:

```c++
std::unique_ptr<CBlockTemplate> CreateNewBlock()
{
    return CreateNewBlock(m_options.coinbase_output_script ? *m_options.coinbase_output_script : CScript{} << OP_TRUE);
}

std::unique_ptr<CBlockTemplate> CreateNewBlock(const CScript& scriptPubKeyIn);
```

So test code in could updated in a separate the commit and then the overload could be deleted.

Sjors replied:
I made an overload, though slightly differently.

ryanofsky reviewed (approved):
Code review ACK 1dfc54d87c7b35cc93cfb81cc2a46c9e8eae7194. Left a style suggestion, and a suggestion to split up the first commit, if we want to do that. But these can be ignored and this all looks good as-is. Makes a lot of sense to not make a parameter which should only be used for testing a required argument.

Sjors commented:
Rebased after #31122.

I changed `coinbase_output_script` from `std::optional` to having a default of `OP_TRUE`.

I split the main commit into three parts:
1. Remove `scriptPubKeyIn` arg from `CreateNewBlock`, using a temporary overload to avoid test changes
2. Dropping the temp overload and change the tests
3. Change the interface

TheCharlatan commented on src/node/types.h:
```diff
@@ -43,6 +44,22 @@ struct BlockCreateOptions {
+     *
+     * Note that higher level code like the getblocktemplate RPC may omit the
+     * coinbase transaction entirely. It's instead constructed by pool software
+     * using fields like coinbaseaux, coinbasevalue, coinbaseaux and
```
Nit: Is there one `coinbaseaux` too many here?

Sjors committed:
rpc: rename coinbase_script to coinbase_output_script

TheCharlatan commented on src/rpc/mining.cpp:
```diff
@@ -814,11 +814,9 @@ static RPCHelpMan getblocktemplate()
         // Create new block
-        CScript scriptDummy = CScript() << OP_TRUE;
-        block_template = miner.createNewBlock(scriptDummy);
+        block_template = miner.createNewBlock(CScript() << OP_TRUE);
```
In commit b35b23d823ac0fc43e4d137525167a7788ad672a:

Nit: I think we just directly call the new function here?

Sjors replied:
This change was premature since the interface is changed in a later commit. It also doesn't do anything, so I'm dropping it.

Sjors commented:
I also split out the renames in 55f2bb7a96806d33e475122f4d8f54b117838afe. They were an accidental find & replace, but useful to be consistent with the new `BlockCreateOptions` field.

Sjors committed:
Drop script_pub_key arg from createNewBlock

Providing a script for the coinbase transaction is only done in test code
and for CPU solo mining.

Production miners use the getblocktemplate RPC which omits the coinbase
transaction entirely from its block template, leaving it to external (pool)
software to construct it.

A coinbase script can still be passed via BlockCreateOptions instead.

A temporary overload is added so that the test can be modified in the
next commit.

Sjors commented:
I re-combined d31e5f9d9986ce1e80b014a1585ab0e944ebcc86 and d1972d751fa882960375050bfa7fc6f4043e0be4 so that the interface change is done along with the change to CreateNewBlock. This seems more clear than having them separate.

Sjors committed:
test: drop scriptPubKeyIn arg from CreateNewBlock

This removes the temporary overload added in the previous commit.

Also drop unneeded custom coinbase output scripts.

Sjors commented:
I dropped a few spurious `(...)coinbase(_output)_script_pub_key` renames from the test commit.

And dropped another unneeded `OP_TRUE` from tx_pool.

sedited reviewed (approved):
ACK 027ce3e9634b2f47c508899942d05690572de516

ryanofsky reviewed (approved):
Code review ACK 027ce3e9634b2f47c508899942d05690572de516. Since last review, this was rebased to avoid conflicts, and option is now a plain cscript instead of an optional\<cscript>, and the commits are split up and there are some minor changes in tests (renames, whitespace, dropping redundant assignments. This is pretty easy to review now that the main commit is so small and changes are split up

Sjors commented:
This PR incidentally "fixes" the IPC bug discovered in https://github.com/bitcoin/bitcoin/pull/31318.

Sjors commented:
Rebased after https://github.com/bitcoin/bitcoin/pull/31112.

ryanofsky reviewed (commented):
Code review ACK 52fd1511a774d1ff9e747b2ce88aa1fcb778ced8. No change since last review other than rebase

sedited reviewed (approved):
Re-ACK 52fd1511a774d1ff9e747b2ce88aa1fcb778ced8

vasild reviewed (approved):
ACK 52fd1511a774d1ff9e747b2ce88aa1fcb778ced8

Looks like a nice cleanup of the interface.

ryanofsky commented:
Merged this but it may be good to address typo vasild found https://github.com/bitcoin/bitcoin/pull/31318#discussion_r1886859006 in another PR

Sjors commented:
Followups in #31520
